### DISCUSSION TITLE
Mi band 4 and room assistant

### DISCUSSION THREAD

**Discussion by barbiermathieu91** [mKeRix/room-assistant#603] [Q&A]
Hello, thank you for this great piece of software.

My basic setup is 3 pri (a raspberry 4 and 2 zeroes) to track a mi band 4

The room tracking is more or less ok but room are often switching ( My wife was really happy to have light turning on at 01:00 am) 

I have tried to to the trilaterion in this post (https://community.home-assistant.io/t/internal-position-tracking-with-co-ordinates-using-an-iphone-and-room-assistant/254399/6) but returned values were weird.

Looking at the RSSI values they seams to be low -72 when I am close and -77~-83 when I am further away and switching frequently.

Does someone has some recommendation to configure room assistant with a mi Band 4? or some recommendation for a smartwatch that could provide better result ?

I plan to play with 'behavior'  but I feel it is gonna sacrifice reactivity

**Comment by mKeRix** [comment]
I don't own a Mi Band 4, so unfortunately I can't give specific input on that, but I can state a few other things. In general I'd recommend to not use Bluetooth as a trigger for turning lights on, as the signal can be affected by a number of different things - leading to the jumping you've observed. Instead, I would recommend to use e.g. a PIR sensor to turn the lights on and then the Bluetooth sensors to determine if the lights should still stay on after some time has passed. You can get these PIRs for quite cheap and even use them with room-assistant directly if you want, just connect them and configure the GPIO integration. This will give you instant triggers for stuff like lights, while counteracting the shortcomings of motion sensors by also letting you detect people that are not moving.

As far as entity behaviors go you are correct that for your problem you'd have to sacrifice reactivity for more stability.

**Reply by barbiermathieu91** [reply]
Ok thanks that seams a more reasonable installation. I have a PIR sensor that I will integrate on the setup.
I have tried another path that look promising. Because the problem is the noisy measurement that makes minimum distance and trialteration difficult, I have tried machine learning. Here what I have done.
I added some esp32 with esphome and the ble_rssi component to add more measurements to my setup. At this point even a fusion + something did not give a satisfactory result.
I have recorded myself for 10 mins in various position/activities in each room with the bracelet and labelled this data with the true Room I was in.
I have trained a random forest classifier with 10 trees and depth 7 and obtain an accuracy above 0.98 and more with higher parameters.
In my opinion this looks promising I will share a little bit more when home. I still need to find a way to include the classifier with home assitant and see the result

**Reply by mKeRix** [reply]
Machine learning is actually something that I want to integrate into room-assistant to make that an easy got use option, as I also think that it can greatly improve tracking for many setups. Until then though, v2.15.1 contains a fix that should make BLE accuracy much better (or rather, more stable). I'd definitely be interested in more details on your setup now.

**Reply by jacopomaroli** [reply]
I started playing with something similar. @barbiermathieu91 can you confirm if you were doing something like this? https://github.com/jacopomaroli/roombutler/blob/master/roombutler/main.py#L208-L232

I just wrapped [this tutorial](https://www.datacamp.com/tutorial/random-forests-classifier-python) into a REST/ws service.
I was thinking to consume ws events from room-assistant and write directly to MQTT (OFC I'll have to disable home-assistant core plugin)

edit: BTW results are a bit underwhelming ATM. probably due to the small sample size, since I trained the whole thing with about ~50 samples per room

**Reply by barbiermathieu91** [reply]
For me the offline result looked great, but after implementing the full system it turns out really janky. I think that bluetooth signals are really affected by lot of external factor orientation people etc. Also my implementation was a bit weird as I tried to add esphome ble implementation. And only using python for the implementation
Turns out I am now doing something related in my work, I will see if I can improve something 
I will look at your project
